takanon_extractions/dynamic_extractions/process_document.py

```diff
@@ -0,0 +1,466 @@
+        
+        # Check if output already exists
+        if self.config.structure_file.exists() and not self.config.overwrite_existing:
+            logger.warning(f"Structure file already exists: {self.config.structure_file}")
+            if not self._confirm_overwrite():
+                logger.info("Skipping structure extraction")
+                self.metadata.stages_completed.append(PipelineStage.EXTRACT_STRUCTURE)
+                return True
+        
+        # Build command
+        cmd = [
+            sys.executable, "extract_structure.py",
+            str(self.config.input_html_file),
+            str(self.config.structure_file),
+            "--environment", self.config.environment.value,
+            "--model", self.config.model,
+            "--max-tokens", str(self.config.max_tokens),
+            "--mark-type", self.config.content_type,
+        ]
+        
+        if self.config.pretty_json:
+            cmd.append("--pretty")
+        
+        # Run command
+        try:
+            logger.info(f"Running command: {' '.join(cmd)}")
+            result = subprocess.run(cmd, capture_output=True, text=True, cwd=Path(__file__).parent)
+            
+            if result.returncode != 0:
+                logger.error(f"Structure extraction failed with return code {result.returncode}")
+                logger.error(f"STDOUT: {result.stdout}")
+                logger.error(f"STDERR: {result.stderr}")
+                self.metadata.errors.append(f"Structure extraction failed: {result.stderr}")
+                return False
+            
+            # Validate output
+            validation_errors = validate_json_structure(
+                self.config.structure_file, 
+                ["metadata", "structure"]
+            )
+            
+            if validation_errors:
+                logger.error("Structure file validation failed:")
+                for error in validation_errors:
+                    logger.error(f"  - {error}")
+                self.metadata.errors.extend(validation_errors)
+                return False
+            
+            # Record stage completion
+            stage_duration = time.time() - stage_start
+            self.metadata.structure_extraction = {
+                "duration_seconds": round(stage_duration, 2),
+                "output_file": str(self.config.structure_file),
+                "file_size_bytes": self.config.structure_file.stat().st_size,
+            }
+            
+            self.metadata.stages_completed.append(PipelineStage.EXTRACT_STRUCTURE)
+            logger.info(f"Structure extraction completed in {stage_duration:.2f}s")
+            
+            return True
+            
+        except Exception as e:
+            logger.error(f"Structure extraction failed with exception: {e}")
+            self.metadata.errors.append(f"Structure extraction exception: {str(e)}")
+            return False
+    
+    def _run_content_extraction(self) -> bool:
```

Comment:
These methods contain lots of duplicated boilerplate - find a way to make this code more compact please.

Reply:
fixed (also added md generation step validation)